File: Competitive Programming 3/Section 1.4/Harder (mode tedious)/220 - Othello/Java/Main.java
import java.util.ArrayList;
import java.util.Scanner;

public class Main{
    static int[][] movs={{1,0},{0,1},{-1,0},{0,-1},{-1,-1},{1,1},{-1,1},{1,-1}};
    static char[][] grid=new char[8][8];
    static boolean inLimit(int i,int j){
        return i>-1&&j>-1&&i<8&&j<8;
    }
    static class brack{
        public int[] mov;
        public int times;
        public brack(int[]mov,int times){
            this.mov=mov;
            this.times=times;
        }
    }
    static int calcDif(int i1,int i2,int j1,int j2){
        return Math.max(Math.abs(j2-j1),Math.abs(i2-i1));
    }
    static int[] possibleLimitCoordinate(int[]mov,int i,int j,char currentPlayer){
        //we are sure that i,j is empty when calling this function
        int iN=i+mov[0],jN=j+mov[1];
        char nextP=nextPlayer(currentPlayer);
        while(inLimit(iN,jN)&&grid[iN][jN]==nextP){
            iN+=mov[0];
            jN+=mov[1];
        }
        if(inLimit(iN,jN)&&grid[iN][jN]==currentPlayer&&calcDif(i,iN,j,jN)>1)
            return new int[]{iN,jN};
        return new int[]{};
    }
    static void listPossibleMoves(char currentPlayer){
        boolean atLeastOne=false;
        for(int i=0;i<8;i++)
            for(int j=0;j<8;j++)
                if(grid[i][j]=='-')
                    for(int mov=0;mov<8;mov++)
                        if(possibleLimitCoordinate(movs[mov],i,j,currentPlayer).length>0){
                            System.out.printf("%s(%d,%d)",atLeastOne?" ":"",i+1,j+1);
                            atLeastOne=true;
                            break;
                        }
        System.out.printf("%s\n",atLeastOne?"":"No legal move.");
    }
    static ArrayList<brack> getBracketed(char currentPlayer,int i,int j){
        ArrayList<brack> bracketed=new ArrayList<>();
        for(int mov=0;mov<8;mov++){
            int[] limitCoord=possibleLimitCoordinate(movs[mov],i,j,currentPlayer);
            if(limitCoord.length!=0)
                bracketed.add(new brack(movs[mov],calcDif(i,limitCoord[0],j,limitCoord[1])-1));
        }
        return bracketed;
    }
    static char nextPlayer(char currenPlayer){
        return currenPlayer=='B'?'W':'B';
    }
    static void makeMoveAndPrintPieceNumbers(int i,int j,char currentPlayer,ArrayList<brack>bracketed){
        grid[i][j]=currentPlayer;
        //brackented required
        for(brack bracket:bracketed){
            int subI=i,subJ=j;
            for (int t=0;t<bracket.times;t++){
                subI+=bracket.mov[0];
                subJ+=bracket.mov[1];
                grid[subI][subJ]=currentPlayer;
            }
        }
        //count
        int black=0,white=0;
        for(int iG=0;iG<8;iG++){
            for(int jG=0;jG<8;jG++){
                if(grid[iG][jG]=='W')
                    white++;
                else if(grid[iG][jG]=='B')
                    black++;
            }
        }
        System.out.printf("Black - %s%d White - %s%d\n",black<10?" ":"",black,white<10?" ":"",white);
    }
    static Scanner s = new Scanner(System.in);
    public static void main(String[]args){
        int t=Integer.parseInt(s.nextLine());
        for(int cs=0;cs<t;cs++){
            if(cs>0)
                System.out.printf("\n");
            for(int i=0;i<8;i++){
                String line=s.nextLine();
                for(int j=0;j<8;j++)
                    grid[i][j]=line.charAt(j);
            }
            char currentPlayer=s.nextLine().charAt(0);
            while(true){
                String command=s.nextLine();
                if(command.charAt(0)=='L')
                    listPossibleMoves(currentPlayer);
                else if(command.charAt(0)=='M'){
                    int i=command.charAt(1)-'0'-1,j=command.charAt(2)-'0'-1;
                    ArrayList<brack>bracketed=getBracketed(currentPlayer,i,j);
                    if(bracketed.size()==0){
                        currentPlayer=nextPlayer(currentPlayer);
                        bracketed=getBracketed(currentPlayer,i,j);
                    }
                    makeMoveAndPrintPieceNumbers(i,j,currentPlayer,bracketed);
                    currentPlayer=nextPlayer(currentPlayer);
                }else{
                    //print grid
                    for(int i=0;i<8;i++){
                        for(int j=0;j<8;j++)
                            System.out.printf("%c",grid[i][j]);
                        System.out.printf("\n");
                    }
                    break;
                }
            }
        }
    }
}

File: Competitive Programming 3/Section 1.4/Harder (mode tedious)/220 - Othello/Python/main.py
movs=[[1,0],[0,1],[-1,0],[0,-1],[1,1],[-1,-1],[-1,1],[1,-1]]
def inLimit(i,j):
    return i>-1 and j>-1 and i < 8 and j < 8
def calcDif(i1,i2,j1,j2):
    return max(abs(j2-j1),abs(i2-i1))
def possibleLimitCoordinate(grid,mov,i,j,currentPlayer):# we are sure that i,j is empty when calling this function
    iN,jN=i+mov[0],j+mov[1]
    nextP=nextPlayer(currentPlayer)
    while inLimit(iN,jN) and grid[iN][jN]==nextP:
        iN+=mov[0]
        jN+=mov[1]
    if inLimit(iN,jN) and grid[iN][jN]==currentPlayer and calcDif(i,iN,j,jN)>1:
        return (iN,jN)
    return None
def listPossibleMoves(currentPlayer,grid):
    atLeastOne=False
    for i in range(8):
        for j in range(8):
            if grid[i][j]=='-':
                for mov in movs:
                    if possibleLimitCoordinate(grid,mov,i,j,currentPlayer):
                        print((' ' if atLeastOne else '')+('('+str(i+1)+','+str(j+1)+')'),end='')
                        atLeastOne=True
                        break
    print('' if atLeastOne else 'No legal move.')
class brack:
    def __init__(self,mov,times):
        self.mov=mov
        self.times=times
def getBracketed(currentPlayer,grid,i,j):
    bracketed=[]
    for mov in movs:
        limitCoord=possibleLimitCoordinate(grid,mov,i,j,currentPlayer)
        if limitCoord!=None:
            dif=calcDif(i,limitCoord[0],j,limitCoord[1])-1
            bracketed.append(brack(mov,dif))
    return bracketed
def nextPlayer(currentPlayer):
    return 'W' if currentPlayer=='B' else 'B'
def makeMoveAndPrintPieceNumbers(i,j,currentPlayer,grid,bracketed):
    grid[i][j]=currentPlayer
    #bracket required
    for bracket in bracketed:
        subI,subJ=i,j
        for _ in range(bracket.times):
            subI+=bracket.mov[0]
            subJ+=bracket.mov[1]
            grid[subI][subJ]=currentPlayer
    #count
    black,white=0,0
    for row in grid:
        for val in row:
            if val=='W':
                white+=1
            elif val=='B':
                black+=1
    print('Black - '+(' ' if black<10 else '')+str(black)+' White - '+(' ' if white<10 else '')+str(white))
def main():
    t = int(input())
    for cs in range(t):
        if cs>0:
            print()
        grid=[]
        for i in range(8):
            grid.append(list(input()))
        currentPlayer = input()
        while True:
            command=input()
            if command[0]=='L':
                listPossibleMoves(currentPlayer,grid)
            elif command[0]=='M':
                i,j=ord(command[1])-ord('0')-1,ord(command[2])-ord('0')-1
                bracketed=getBracketed(currentPlayer,grid,i,j)
                if bracketed==[]:
                    currentPlayer=nextPlayer(currentPlayer)
                    bracketed=getBracketed(currentPlayer,grid,i,j)
                makeMoveAndPrintPieceNumbers(i,j,currentPlayer,grid,bracketed)
                currentPlayer=nextPlayer(currentPlayer)
            else:#printgrid
                for row in grid:
                    for val in row:
                        print(val,end='')
                    print()
                break           
main()